mautic/mautic
# Pipedrive merge support

:warning: Do not merge this PR before https://github.com/mautic/mautic/pull/11228 is merged as it is dependent on that PR. ⚠️ 



| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [x]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | 
| Related developer documentation PR URL | 
| Issue(s) addressed                     | 



#### Description:


This PR makes it possible to integrate the "merge" operation between Mautic and Pipedrive for both companies and contacts.

Once configured, merging on the Pipedrive side will cause the corresponding merge to occur natively in Mautic, and starting a merge from the Mautic side will cause the equivalent merge to occur natively in Pipedrive.

Important note 1:
Included in this PR are the commits from another PR, https://github.com/mautic/mautic/pull/11228. They create the Company merge event in Mautic that this PR relies on. Thus, this PR can still be tested, but it should probably not be merged until after that PR is.

Important note 2:
Pipedrive seems to have a quirky implementation of merging webhooks, so this isn't as straightforward as it should be. Upon merging on the Pipedrive side, it dispatches two webhooks: merge (for the updated "winning" record) and delete (for the previous "losing" record). The delete request looks just like a normal delete request unrelated to a merge, so I haven't found a way to tell during the processing of that request that it should be ignored. This wouldn't be a big deal, except that the order the webhooks are received is not guaranteed and in my testing sometimes the delete comes before the merge. This makes it impossible to then complete the merge on the Mautic side if the losing record is already deleted.

To work around this issue, there is an optional feature for the Pipedrive integration to handle deletions with a custom command via a cron task instead. After enabling the setting, you must run the command via cron. Regardless of how often the cron job runs, the site will wait a minimum of 60s before processing the deletion to ensure there is plenty of time for the merge request to complete if there is one.

While this seems to work in testing, I don't like it as a real solution. However, I also haven't found any better way to prevent the deletion from occurring before the merge.

#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create a demo account on Pipedrive (if using an existing account, make sure it is empty or contains only test data)
3. Within Mautic, set up the Pipedrive integration and enter your Pipedrive connection information. Leave the "Handle deletions from Pipedrive via cron" setting disabled for now
4. Within Pipedrive, set up the webhooks for Mautic per the integration's typical installation steps with the addition of also enabling the "merge" webhook
5. Create 3 companies and 3 contacts in Mautic. The records should also appear in Pipedrive. If they do not, fix the integration setup issue and try again
6. Create 2 companies and 2 contacts in Pipedrive, different from the previous ones. The records should also appear in Mautic. If not, fix the integration setup issue and try again
7. Delete one contact and one company from Pipedrive and make sure the record is deleted from Mautic. This test ensures that without the cron setting enabled, the delete webhook from Pipedrive still works like normal
8. Edit the Pipedrive integration in Mautic and enable the "Handle deletions from Pipedrive via cron" setting for the rest of the testing. From now on out, Pipedrive deletions will only happen within Mautic when you run the `mautic:integration:pipedrive:process-deletions` command through the CLI or via a cron task at least 60 seconds after the deletion occurred
9. Merge any two companies in Mautic, and ensure that the companies are properly merged (with all expected values) in Pipedrive
10. Merge any two contacts in Mautic, and ensure that they are properly merged (with expected winning values) in Pipedrive
11. Merge two companies in Pipedrive, and ensure that the companies are properly merged (with all expected values) in Mautic
12. Merge two contacts in Pipedrive, and ensure that they are properly merged (with expected winning values) in Mautic
14. Run the `mautic:integration:pipedrive:process-deletions` CLI command in order to process the previous Pipedrive deletions. Since the records were already merged, they should no longer exist to be deleted, but the command should still succeed in processing the pending records without error




<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/11229"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
